pkg/controllers/endpointaccessible/endpoint_accessible_controller.go

```diff
@@ -0,0 +1,219 @@
+	operatorClient v1helpers.OperatorClient,
+	corev1Informers corev1informers.Interface,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsLister := corev1Informers.Endpoints().Lister()
+	endpointsInformer := corev1Informers.Endpoints().Informer()
+
+	endpointsListFn := func() ([]string, error) {
+		return listOAuthServiceEndpoints(endpointsLister, recorder)
+	}
+
+	return NewEndpointAccessibleController(
+		"OAuthServiceEndpointsCheck",
+		operatorClient, endpointsListFn, []factory.Informer{endpointsInformer}, recorder)
+}
+
+func listOAuthServiceEndpoints(endpointsLister corev1listers.EndpointsLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	endpoints, err := endpointsLister.Endpoints("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthServiceEndpointsCheck", "failed to get oauth service endpoints: %v", err)
+		return results, nil
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				results = append(results, net.JoinHostPort(address.IP, strconv.Itoa(int(port.Port))))
+			}
+		}
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("oauth service endpoints are not ready")
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthServices(serviceLister corev1listers.ServiceLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	service, err := serviceLister.Services("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthServiceCheck", "failed to get oauth service: %v", err)
+		return nil, err
+	}
+	for _, port := range service.Spec.Ports {
+		results = append(results, net.JoinHostPort(service.Spec.ClusterIP, strconv.Itoa(int(port.Port))))
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("no valid oauth services found")
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthRoutes(routeLister routev1listers.RouteLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	route, err := routeLister.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthRouteCheck", "failed to get oauth route: %v", err)
+		return nil, err
+	}
+	for _, ingress := range route.Status.Ingress {
+		if len(ingress.Host) > 0 {
+			results = append(results, ingress.Host)
+		}
+	}
+	if len(results) == 0 {
+		recorder.Warningf("OAuthRouteCheck", "route status does not have host address")
+		return nil, fmt.Errorf("route status does not have host address")
+	}
+	return toHealthzURL(results), nil
+}
+
+type endpointAccessibleController struct {
+	operatorClient v1helpers.OperatorClient
+	endpointListFn EndpointListFunc
+}
+
+type EndpointListFunc func() ([]string, error)
+
+// NewEndpointAccessibleController returns a controller that checks if the endpoints
+// listed by endpointListFn are reachable
+func NewEndpointAccessibleController(
```

Comment:
in a followup PR, I'd like to see this moved to a separate package.  It's the reusable collateral.

Reply:
I made a note of it and will move it to separate package in follow up PR.